library/alloc/src/string.rs

```diff
@@ -2444,6 +2452,49 @@ impl ToString for i8 {
     }
 }
 
+#[cfg(not(no_global_oom_handling))]
+#[stable(feature = "u16_to_string_specialization", since = "0.00.0")]
+impl ToString for u16 {
+    #[inline]
+    fn to_string(&self) -> String {
+        let mut buf = String::with_capacity(u16::MAX.log10() as usize + 1);
```

Comment:
As mentioned by [MIN_NON_ZERO_CAP](https://github.com/rust-lang/rust/blob/1067e2ca5e9cfe5c79f956e49ffc684c5142d49b/library/alloc/src/raw_vec.rs#L120), it is very unlikely you will see any benefit from allocating 5 or 6 bytes instead of 8. Allocators don't really work on a byte level. They have certain internal alignments and granularities, regardless of the size/alignment being requested.

Reply:
Interesting. Can you link some source?

Reply:
Well generally OS memory allocators are designed to support malloc, which has requirements such as `The address of a block returned by malloc or realloc in GNU systems is always a multiple of eight (or sixteen on 64-bit systems)` (https://www.gnu.org/software/libc/manual/html_node/Aligned-Memory-Blocks.html), You can also see the same alignment requirements by rust https://github.com/rust-lang/rust/blob/master/library/std/src/sys/common/alloc.rs
Although pointer alignment doesn't necessarily equal size alignment, it doesn't really make sense for these allocators to bother supporting sizes less than the minimum alignment.

The best sources I know off the top of my head are just how some popular allocators work and you will see they group small allocations into certain sizes, usually multiples of 8 or 16.
https://github.com/jemalloc/jemalloc/blob/12cd13cd418512d9e7596921ccdb62e25a103f87/include/jemalloc/internal/sc.h#L7
https://github.com/microsoft/mimalloc/blob/752594e76423526e108413731518a26e3322b9ca/include/mimalloc-internal.h#L376
https://github.com/oneapi-src/oneTBB/blob/e82d250338d3253b862c96715637870e7cfe408f/src/tbbmalloc/frontend.cpp#L802